File: command_center/renderers/table_renderer.py
"""
Command Center — Table Renderer
==================================
Produces structured table blocks from data.
"""

from typing import Any, Dict, List, Optional


def create_table_block(
    headers: List[str],
    rows: List[List[Any]],
    title: Optional[str] = None,
) -> Dict[str, Any]:
    """Create a table content block."""
    block = {
        "type": "table",
        "headers": headers,
        "rows": rows,
    }
    if title:
        block["title"] = title
    return block


def dict_list_to_table(data: List[Dict[str, Any]], title: Optional[str] = None) -> Dict[str, Any]:
    """Convert a list of dicts to a table block."""
    if not data:
        return create_table_block([], [], title)

    headers = list(data[0].keys())
    rows = [[str(row.get(h, "")) for h in headers] for row in data]
    return create_table_block(headers, rows, title)

File: builder_service/static/js/table-renderer.js
/**
 * Table Renderer
 * Creates styled HTML tables from structured block data.
 */

export class TableRenderer {
    /**
     * Create a table element from a block definition.
     * @param {object} block — table block data
     * @returns {HTMLElement} the table wrapper element
     */
    create(block) {
        const wrapper = document.createElement('div');
        wrapper.className = 'block-table';

        // Title bar
        if (block.title) {
            const titleBar = document.createElement('div');
            titleBar.className = 'block-table-title';
            titleBar.textContent = block.title;
            wrapper.appendChild(titleBar);
        }

        // Scroll container for wide tables
        const scrollWrap = document.createElement('div');
        scrollWrap.className = 'block-table-scroll';

        const table = document.createElement('table');

        // Header
        if (block.headers && block.headers.length > 0) {
            const thead = document.createElement('thead');
            const tr = document.createElement('tr');
            for (const header of block.headers) {
                const th = document.createElement('th');
                th.textContent = header;
                tr.appendChild(th);
            }
            thead.appendChild(tr);
            table.appendChild(thead);
        }

        // Body
        if (block.rows && block.rows.length > 0) {
            const tbody = document.createElement('tbody');
            for (const row of block.rows) {
                const tr = document.createElement('tr');
                for (const cell of row) {
                    const td = document.createElement('td');
                    td.textContent = String(cell ?? '');
                    tr.appendChild(td);
                }
                tbody.appendChild(tr);
            }
            table.appendChild(tbody);
        }

        scrollWrap.appendChild(table);
        wrapper.appendChild(scrollWrap);

        return wrapper;
    }
}
